flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/spec/DynamicTableSourceSpec.java

```diff
@@ -95,29 +92,19 @@ private DynamicTableSource getTableSource(FlinkContext context, FlinkTypeFactory
                     FactoryUtil.createDynamicTableSource(
                             factory,
                             contextResolvedTable.getIdentifier(),
-                            resolvedCatalogTable,
+                            contextResolvedTable.getResolvedTable(),
                             loadOptionsFromCatalogTable(contextResolvedTable, context),
                             context.getTableConfig(),
                             context.getClassLoader(),
                             contextResolvedTable.isTemporary());
-            // validate DynamicSource and apply Metadata
-            DynamicSourceUtils.prepareDynamicSource(
-                    contextResolvedTable.getIdentifier().toString(),
-                    resolvedCatalogTable,
-                    tableSource,
-                    false,
-                    context.getTableConfig().getConfiguration());
 
             if (sourceAbilities != null) {
-                //  Note: use DynamicSourceUtils.createProducedType to produce the type info so that
-                //  keep consistent with sql2Rel phase which also called the method producing
-                //  deterministic format (PHYSICAL COLUMNS + METADATA COLUMNS) when converts a given
-                //  DynamicTableSource to a RelNode.
-                // TODO should do a refactor(e.g., add serialized input type info into each
-                //  SourceAbilitySpec so as to avoid this implicit logic dependency)
                 RowType newProducedType =
-                        DynamicSourceUtils.createProducedType(
-                                contextResolvedTable.getResolvedSchema(), tableSource);
+                        (RowType)
+                                contextResolvedTable
+                                        .getResolvedSchema()
+                                        .toSourceRowDataType()
+                                        .getLogicalType();
```

Comment:
Hi, @dawidwys, thanks for your PR. I have a small question that why we change the spec column index from `PHYSICAL COLUMNS + METADATA COLUMNS` to the original user-defined schema order here ? If we want to change this order, more code that uses source specs need to be modified, such as `ScanReuser.reuseDuplicatedScan`, cc @tillrohrmann .

Reply:
cc @twalthr

Reply:
Not sure, if I understand your question. 

Before applying any specs the table has a schema as described by 
```
contextResolvedTable.getResolvedSchema()
                .toSourceRowDataType()
                .getLogicalType()
```

Only once the `ReadingMetadataSpec` is applied it will change the type to `spec.getProducedType().get()`. In this particular case of reading metadata in a `TableSourceScan` it is equal to `DynamicSourceUtils.createProducedType(contextResolvedTable.getResolvedSchema(), tableSource);`

Reply:
Hi, @dawidwys . For this case:
```
String ddl =  "CREATE TABLE MyTable1 (\n"
                        + "  metadata_0 int METADATA VIRTUAL,\n"
                        + "  a0 int,\n"
                        + "  a1 int,\n"
                        + "  a2 int,\n"
                        + "  ts STRING,\n "
                        + "  rowtime as TO_TIMESTAMP(`ts`),\n"
                        + "  WATERMARK FOR rowtime AS rowtime - INTERVAL '1' SECOND\n"
                        + ") WITH ( .....
                        + ")";
```
which `METADATA` column defined as first column.  
```
DynamicSourceUtils.createProducedType(contextResolvedTable.getResolvedSchema(), tableSource)
``` 
will get different schema order with
```
contextResolvedTable.getResolvedSchema()
                .toSourceRowDataType()
                .getLogicalType()
```

IMO, I think it's better to use the origin order than convert the order to `PHYSICAL COLUMNS + METADATA COLUMNS`, but this need @twalthr to confirm it.

Reply:
By just looking at the code, I would say `newProducedType` should be initialized with only physical columns. Not sure what the tests say to this. But `toSourceRowDataType` includes also computed column which makes no sense for the DynamicTableSource as the source doesn't know this concept. After the spec is applied in the loop, the type changes for `PHYSICAL COLUMNS + METADATA COLUMNS`. In any case, the producedType should always contain what comes out of the source.

Reply:
I updated the PR to use `toPhysicalRowDataType` as the initial type.